File: src/main/java/org/hisp/dhis/rules/RuleEngineExecution.java
package org.hisp.dhis.rules;

import org.hisp.dhis.rules.models.Rule;
import org.hisp.dhis.rules.models.RuleEffect;
import org.hisp.dhis.rules.models.RuleEnrollment;
import org.hisp.dhis.rules.models.RuleEvent;
import org.hisp.dhis.rules.models.TrackerObjectType;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

record RuleEngineExecution(
        @CheckForNull RuleEvent event,
        @CheckForNull RuleEnrollment enrollment,
        @Nonnull Map<String, List<String>> supplementaryData,
        @Nonnull List<Rule> rules,
        @Nonnull Map<String, RuleVariableValue> valueMap,
        @Nonnull RuleConditionEvaluator ruleConditionEvaluator
) implements Callable<List<RuleEffect>> {

    RuleEngineExecution(@Nonnull RuleEvent event, @Nonnull List<Rule> rules,
                        @Nonnull Map<String, RuleVariableValue> valueMap, @Nonnull Map<String, List<String>> supplementaryData) {
        this(event, null, supplementaryData, rules, new HashMap<>(valueMap), new RuleConditionEvaluator());
    }

    RuleEngineExecution(@Nonnull RuleEnrollment enrollment, @Nonnull List<Rule> rules,
                        @Nonnull Map<String, RuleVariableValue> valueMap, @Nonnull Map<String, List<String>> supplementaryData) {
        this(null, enrollment, supplementaryData, rules, new HashMap<>(valueMap), new RuleConditionEvaluator());
    }

    @Override
    public List<RuleEffect> call() {
        if (event != null) {
            return ruleConditionEvaluator.getRuleEffects(TrackerObjectType.EVENT, event.event(), valueMap,
                    supplementaryData, this.rules);
        }
        if (enrollment != null) {
            return ruleConditionEvaluator.getRuleEffects(TrackerObjectType.ENROLLMENT, enrollment.getEnrollment(), valueMap,
                    supplementaryData, this.rules);
        }
        throw new IllegalStateException("event and enrollment were null");
    }
}

File: src/jvmMain/kotlin/org/hisp/dhis/rules/RuleEngineExecution.kt
package org.hisp.dhis.rules

import org.hisp.dhis.rules.models.*
import java.util.concurrent.Callable


internal data class RuleEngineExecution(
    val event: RuleEvent?,
    val enrollment: RuleEnrollment?,
    val rules: List<Rule>,
    val valueMap: MutableMap<String, RuleVariableValue>,
    val supplementaryData: Map<String, List<String>>
) : Callable<List<RuleEffect>> {
    override fun call(): List<RuleEffect> {
        if (event != null) {
            return RuleConditionEvaluator().getRuleEffects(
                TrackerObjectType.EVENT, event.event(), valueMap,
                supplementaryData, rules
            )
        }
        if (enrollment != null) {
            return RuleConditionEvaluator().getRuleEffects(
                TrackerObjectType.ENROLLMENT, enrollment.enrollment, valueMap,
                supplementaryData, rules
            )
        }
        throw IllegalStateException("event and enrollment were null")
    }
}
